File: 250713/0048/minkyu.cpp
#include <bits/stdc++.h>
#include <iterator>

using namespace std;

class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int two_c = n - 1;
        int x_max = (int) ceil(n / 2.0);
        int y_max = (int) ceil((n - 1) / 2.0);
        int old_x, old_y, new_x, new_y, old_value;
        for (int x = 0; x < x_max; x++) {
            for (int y = 0; y < y_max; y++) {
                old_value = matrix[x][y];
                old_x = x;
                old_y = y;
                for (int i = 0; i < 4; i++) {
                    new_x = old_y;
                    new_y = two_c - old_x;
                    swap(old_value, matrix[new_x][new_y]);
                    old_x = new_x;
                    old_y = new_y;
                }
            }
        }
    }

    void rotate2(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }

        for (int i = 0; i < n; i++) {
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};

File: 250713/0048/minkyu.py
import math
from typing import List

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        two_c = n - 1
        x_max = math.ceil(n / 2)
        y_max = math.ceil((n - 1) / 2)
        for x in range(x_max):
            for y in range(y_max):
                old_value = matrix[x][y]
                old_x = x
                old_y = y
                for _ in range(4):
                    """
                    Let c be the coorinate of the center of the matrix.
                    e.g. c = 1 when n = 3, c = 1.5 when n = 4
                    original coordinate: x + yi
                    after translation: (x + yi) - (c + ci) = (x - c) + (y - c)i
                    after rotation: ((x - c) + (y - c)i) * -i = (y - c) + (c - x)i
                    after recovery of translation: (y - c) + (c - x)i + (c + ci) = y + (2c - x)i
                    """
                    new_x = old_y
                    new_y = two_c - old_x
                    temp = matrix[new_x][new_y]
                    matrix[new_x][new_y] = old_value
                    old_value = temp
                    old_x = new_x
                    old_y = new_y

    def rotate2(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        for i in range(n):
            for j in range(i + 1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

        for i in range(int(n / 2)):
            matrix[i], matrix[n - i] = matrix[n - i], matrix[i]
